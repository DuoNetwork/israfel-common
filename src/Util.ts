import moment, { DurationInputArg2 } from 'moment';
import WebSocket from 'ws';
import * as CST from './constants';

export class Util {
	public static logLevel: string = CST.LOG_INFO;

	public static logInfo(text: any): void {
		this.log(text, CST.LOG_INFO);
	}

	public static logDebug(text: any): void {
		this.log(text, CST.LOG_DEBUG);
	}

	public static logError(text: any): void {
		this.log(text, CST.LOG_ERROR);
	}

	private static log(text: any, level: string) {
		if (CST.LOG_RANKING[this.logLevel] >= CST.LOG_RANKING[level])
			console.log(
				`${moment.utc(Util.getUTCNowTimestamp()).format('DD HH:mm:ss.SSS')} [${level}]: ` +
					text
			);
	}

	public static isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public static isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}

	public static getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public static round(num: string | number) {
		return +(Math.round((num + 'e+8') as any) + 'e-8');
	}

	public static safeWsSend(ws: WebSocket, message: string) {
		try {
			ws.send(message);
			return true;
		} catch (error) {
			this.logError(error);
			return false;
		}
	}

	public static sleep(ms: number) {
		return new Promise(resolve => {
			global.setTimeout(resolve, ms);
		});
	}

	public static clone(obj: object) {
		return JSON.parse(JSON.stringify(obj));
	}

	public static getDates(
		length: number,
		step: number,
		stepSize: DurationInputArg2,
		format: string
	) {
		const dates: string[] = [];
		const date = moment.utc(this.getUTCNowTimestamp());
		for (let i = 0; i < length; i++) {
			dates.push(date.format(format));
			date.subtract(step, stepSize);
		}
		dates.sort((a, b) => a.localeCompare(b));

		return dates;
	}

	public static getMonthEndExpiry(timestamp: number) {
		const dateObj = moment
			.utc(timestamp)
			.endOf('month')
			.startOf('day');
		const day = dateObj.day();
		if (day === 6) dateObj.subtract(1, 'day');
		else if (day < 5) dateObj.subtract(day + 2, 'day');

		dateObj.add(8, 'hour');

		return dateObj.valueOf();
	}

	public static getDayExpiry(timestamp: number) {
		const dateObj = moment.utc(timestamp).startOf('day');
		dateObj.add(8, 'hour');

		return dateObj.valueOf();
	}

	public static getExpiryTimestamp(isMonth: boolean) {
		const now = this.getUTCNowTimestamp();
		if (isMonth) {
			const thisMonthEndExpiry = this.getMonthEndExpiry(now);
			if (now > thisMonthEndExpiry - 4 * 3600000)
				return this.getMonthEndExpiry(
					moment
						.utc(thisMonthEndExpiry)
						.add(1, 'week')
						.valueOf()
				);
			return thisMonthEndExpiry;
		} else {
			const todayExpiry = this.getDayExpiry(now);
			if (now > todayExpiry - 4 * 3600000) return todayExpiry + 24 * 3600000;
			return todayExpiry;
		}
	}

	public static formatFixedNumber(num: number, precision: number) {
		const decimal = precision && precision < 1 ? (precision + '').length - 2 : 0;
		const roundedNumber = Math.round(num / precision) * precision;
		return precision ? roundedNumber.toFixed(decimal) : num + '';
	}
}

export default Util;
