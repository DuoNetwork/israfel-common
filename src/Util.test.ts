import * as CST from './constants';
import Util from './Util';

test('isNumber() return true for numbers', () => {
	expect(Util.isNumber(5)).toBe(true);
	expect(Util.isNumber(5.0)).toBe(true);
});

test('isNumber() return true for empty string and null', () => {
	expect(Util.isNumber('')).toBe(true);
	expect(Util.isNumber(null)).toBe(true);
});

test('isNumber() return true for number strings', () => {
	expect(Util.isNumber('5')).toBe(true);
	expect(Util.isNumber('5.0')).toBe(true);
});

test('isNumber() return false for other strings', () => {
	expect(Util.isNumber('5.0s')).toBe(false);
	expect(Util.isNumber('test')).toBe(false);
	expect(Util.isNumber('NaN')).toBe(false);
});

test('isNumber() return false for undefined, infinity, NaN', () => {
	expect(Util.isNumber(undefined)).toBe(false);
	expect(Util.isNumber(Infinity)).toBe(false);
	expect(Util.isNumber(NaN)).toBe(false);
});

test('{}, null, undefined is empty', () => {
	expect(Util.isEmptyObject({})).toBe(true);
	expect(Util.isEmptyObject(null)).toBe(true);
	expect(Util.isEmptyObject(undefined)).toBe(true);
});

test('{test: true} is not empty', () => {
	expect(Util.isEmptyObject({ test: true })).toBe(false);
});

test('round', () => {
	expect(Util.round('12345')).toMatchSnapshot();
	expect(Util.round('12345.000')).toMatchSnapshot();
	expect(Util.round('12345.1234567')).toMatchSnapshot();
	expect(Util.round('12345.123456789')).toMatchSnapshot();
	expect(Util.round('0.123456789123456789')).toMatchSnapshot();
	expect(Util.round('12345.123456789123456789')).toMatchSnapshot();
});

test('safeWsSend', () => {
	const ws = {
		send: jest.fn()
	};
	expect(Util.safeWsSend(ws as any, 'message')).toBeTruthy();
	ws.send = jest.fn(() => {
		throw new Error('error');
	});
	expect(Util.safeWsSend(ws as any, 'message')).toBeFalsy();
});

test('getDates', () => {
	Util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	expect(Util.getDates(4, 1, 'days', 'YYYY-MM-DD')).toMatchSnapshot();
});

test('getExpiryTimeStamp', () => {
	Util.getUTCNowTimestamp = jest.fn(() => 1544519089000); // 2018-12-11 9:04 UTC
	expect(Util.getExpiryTimestamp(false)).toBe(1544601600000); // 2018-12-12 8:00 UTC
	expect(Util.getExpiryTimestamp(true)).toBe(1544774400000); // 2018-12-14 8:00 UTC
	Util.getUTCNowTimestamp = jest.fn(() => 1544493600000); // 2018-12-11 2:00 UTC
	expect(Util.getExpiryTimestamp(false)).toBe(1544515200000); // 2018-12-11 8:00 UTC
	Util.getUTCNowTimestamp = jest.fn(() => 1544760000000); // 2018-12-14 4:00 UTC
	expect(Util.getExpiryTimestamp(true)).toBe(1544774400000); // 2018-12-14 8:00 UTC
	Util.getUTCNowTimestamp = jest.fn(() => 1544760000001); // 2018-12-14 4:00:00.01 UTC
	expect(Util.getExpiryTimestamp(true)).toBe(1545379200000); // 2018-12-21 8:00 UTC
	Util.getUTCNowTimestamp = jest.fn(() => 1544832000000); // 2018-12-15 0:00 UTC
	expect(Util.getExpiryTimestamp(true)).toBe(1545379200000); // 2018-12-21 8:00 UTC
});

test('sleep', async () => {
	global.setTimeout = jest.fn(resolve => resolve()) as any;
	await Util.sleep(1);
	expect((global.setTimeout as jest.Mock).mock.calls).toMatchSnapshot();
});

test('formatFixedNumber', () => {
	expect(Util.formatFixedNumber(123.456789, 0)).toBe('123.456789');
	expect(Util.formatFixedNumber(123.456789, 0.5)).toBe('123.5');
});

test('log debug', () => {
	Util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	console.log = jest.fn();

	Util.logLevel = CST.LOG_DEBUG;
	Util.logError('error');
	Util.logInfo('info');
	Util.logDebug('debug');
	expect((console.log as jest.Mock).mock.calls).toMatchSnapshot();
});

test('log info', () => {
	Util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	console.log = jest.fn();

	Util.logLevel = CST.LOG_INFO;
	Util.logError('error');
	Util.logInfo('info');
	Util.logDebug('debug');
	expect((console.log as jest.Mock).mock.calls).toMatchSnapshot();
});

test('log error', () => {
	Util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	console.log = jest.fn();

	Util.logLevel = CST.LOG_ERROR;
	Util.logError('error');
	Util.logInfo('info');
	Util.logDebug('debug');
	expect((console.log as jest.Mock).mock.calls).toMatchSnapshot();
});
