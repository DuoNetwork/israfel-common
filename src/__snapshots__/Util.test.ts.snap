// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDates 1`] = `
Array [
  "1970-01-12",
  "1970-01-13",
  "1970-01-14",
  "1970-01-15",
]
`;

exports[`log debug 1`] = `
Array [
  Array [
    "15 06:56:07.890 [ERROR]: error",
  ],
  Array [
    "15 06:56:07.890 [INFO]: info",
  ],
  Array [
    "15 06:56:07.890 [DEBUG]: debug",
  ],
]
`;

exports[`log error 1`] = `
Array [
  Array [
    "15 06:56:07.890 [ERROR]: error",
  ],
]
`;

exports[`log info 1`] = `
Array [
  Array [
    "15 06:56:07.890 [ERROR]: error",
  ],
  Array [
    "15 06:56:07.890 [INFO]: info",
  ],
]
`;

exports[`round 1`] = `12345`;

exports[`round 2`] = `12345`;

exports[`round 3`] = `12345.1234567`;

exports[`round 4`] = `12345.12345679`;

exports[`round 5`] = `0.12345679`;

exports[`round 6`] = `12345.12345679`;

exports[`sleep 1`] = `
Array [
  Array [
    [Function],
    1,
  ],
]
`;
